internal/pkg/xds/xds_mcp.go

```diff
@@ -0,0 +1,202 @@
+type (
+	DiscoveryStream = discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer
+	DeltaDiscoveryStream = discovery.AggregatedDiscoveryService_DeltaAggregatedResourcesServer
+)
+
+// adsServer implements Envoy's AggregatedDiscoveryService service for sending MCP resources to Istiod.
+// ads is Aggregated Discovery Service
+type adsServer struct {
+	subscribers      sync.Map
+	nextSubscriberID atomic.Uint64
+}
+
+// subscriber represents a client that is subscribed to MCP resources.
+type subscriber struct {
+	id uint64
+	stream      DiscoveryStream
+	closeStream func()
+}
+
+var _ discovery.AggregatedDiscoveryServiceServer = (*adsServer)(nil)
+
+// NewADSServer creates a new instance of the AggregatedDiscoveryServiceServer.
+func (adss *adsServer) StreamAggregatedResources(downstream DiscoveryStream) error {
+	log.Println("New subscriber connected")
+	ctx, closeStream := context.WithCancel(downstream.Context())
+
+	sub := &subscriber{
+		id:          adss.nextSubscriberID.Add(1),
+		stream:      downstream,
+		closeStream: closeStream,
+	}
+
+	adss.subscribers.Store(sub.id, sub)
+	go recvFromStream(int64(sub.id), downstream)
+
+	<-ctx.Done()
+	return nil
+}
+
+// DeltaAggregatedResources is not implemented.
+func (adss *adsServer) DeltaAggregatedResources(downstream DeltaDiscoveryStream) error {
+	return status.Errorf(codes.Unimplemented, "Not Implemented")
+}
+
+
+var (
+	maxUintDigits = len(strconv.FormatUint(uint64(math.MaxUint64), 10))
+	subIDFmtStr   = `%0` + strconv.Itoa(maxUintDigits) + `d`
+)
+
+// recvFromStream receives discovery requests from the subscriber.
+func recvFromStream(id int64, downstream DiscoveryStream) {
+	log.Println("Received from stream ", id)
+	recvLoop:
+		for {
+			discoveryRequest, err := downstream.Recv()
+			if err != nil {
+				log.Print("Error while recv discovery request from subscriber ", fmt.Sprintf(subIDFmtStr, id), err)
+				break recvLoop
+			}
+			log.Print("Got discovery request from subscriber ", fmt.Sprintf(subIDFmtStr, id), discoveryRequest)
+			if discoveryRequest.GetVersionInfo() == "" {
+				log.Print("Send initial empty config snapshot for type ", discoveryRequest.GetTypeUrl())
+				sendToStream(downstream, discoveryRequest.GetTypeUrl(), make([]*anypb.Any, 0), strconv.FormatInt(time.Now().Unix(), 10))
+			}
+		}
+	}
+
+// sendToStream sends MCP resources to the subscriber.
+func sendToStream(downstream DiscoveryStream, typeUrl string, mcpResources []*anypb.Any, version string) error {
+	if err := downstream.Send(&discovery.DiscoveryResponse{
+		TypeUrl:     typeUrl,
+		VersionInfo: version,
+		Resources:   mcpResources,
+		ControlPlane: &envoycfgcorev3.ControlPlane{
+			Identifier: os.Getenv("POD_NAME"),
+		},
+		Nonce: version,
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// pushToSubscribers pushes MCP resources to active subscribers.
+func (adss *adsServer) pushToSubscribers() error {
+	mcpResources, err := makeMCPResources(numMCPResources)
```

Comment:
I would like to provide some working test data. Could we send the following service entries?
```suggestion
var mcpResources []*anypb.Any
example, err := makeMCPServiceEntry("example.com")
mcpResources = append(mcpResources, example)
istio, err := makeMCPServiceEntry("istio.io")
mcpResources = append(mcpResources, istio)
```